File: server/src/main/java/gg/mineral/server/network/packet/play/clientbound/RespawnPacket.java
package gg.mineral.server.network.packet.play.clientbound;

import gg.mineral.api.entity.living.human.property.Gamemode;
import gg.mineral.api.network.packet.Packet;
import gg.mineral.api.world.property.Difficulty;
import gg.mineral.api.world.property.Dimension;
import gg.mineral.api.world.property.LevelType;
import io.netty.buffer.ByteBuf;

public final record RespawnPacket(Dimension dimension, Difficulty difficulty, Gamemode gamemode, LevelType levelType)
        implements Packet.OUTGOING {
    @Override
    public void serialize(ByteBuf os) {
        os.writeInt(dimension.getId());
        os.writeByte(difficulty.getId());
        os.writeByte(gamemode.getId());
        writeString(os, levelType.string());
    }

    @Override
    public byte getId() {
        return 0x07;
    }
}

File: server/src/main/kotlin/gg/mineral/server/network/packet/play/clientbound/RespawnPacket.kt
package gg.mineral.server.network.packet.play.clientbound

import gg.mineral.api.entity.living.human.property.Gamemode
import gg.mineral.api.network.packet.Packet
import gg.mineral.api.world.property.Difficulty
import gg.mineral.api.world.property.Dimension
import gg.mineral.api.world.property.LevelType
import io.netty.buffer.ByteBuf

@JvmRecord
data class RespawnPacket(
    val dimension: Dimension,
    val difficulty: Difficulty,
    val gamemode: Gamemode,
    val levelType: LevelType
) :
    Packet.OUTGOING {
    override fun serialize(os: ByteBuf) {
        os.writeInt(dimension.id.toInt())
        os.writeByte(difficulty.id.toInt())
        os.writeByte(gamemode.id.toInt())
        writeString(os, levelType.string())
    }

    override val id: Byte
        get() = 0x07
}
